File: composite/composite.cpp
#include <iostream>
#include <list>
#include <memory>

class IShape {
public:
    typedef std::shared_ptr<IShape> SPtr;

    virtual void draw() = 0;

    virtual void addComponent(const SPtr&) {
        throw std::runtime_error("SPtr: Can't add to a leaf");
    }

    virtual void removeComponent(const SPtr&) {
        throw std::runtime_error("SPtr: Can't remove from a leaf");
    }
};

class Square : public IShape {
    virtual void draw() {
        std::cout << "I'm a square" << std::endl;
    }
};

class Circle : public IShape {
    virtual void draw() {
        std::cout << "I'm a circle" << std::endl;
    }
};

class Triangle : public IShape {
    virtual void draw() {
        std::cout << "I'm a triangle" << std::endl;
    }
};

class Point : public IShape {
    virtual void draw() {
        std::cout << "I'm a square" << std::endl;
    }
};

class Composite : public IShape {
public:
    virtual void draw() {
        for (SPtr& shape : components_) {
            shape->draw();
        }
    }

    void addComponent(const SPtr& sptr) {
        components_.push_back(sptr);
    }

    void removeComponent(const SPtr& sptr) {
        components_.remove(sptr);
    }

private:
    std::list <SPtr> components_;
};

int main() {
    std::cout << "Hello, World!" << std::endl;

    Composite comp;

    IShape::SPtr shape1(new Square);
    IShape::SPtr shape2(new Circle);
    IShape::SPtr shape3(new Triangle);
    IShape::SPtr shape4(new Point);

    IShape::SPtr comp2(new Composite);
    IShape::SPtr comp3(new Composite);

    comp2->addComponent(shape1);
    comp2->addComponent(shape2);
    comp2->addComponent(shape3);

    comp3->addComponent(shape4);

    comp.addComponent(comp2);
    comp.addComponent(comp3);

    comp.draw();

    return 0;
}

File: composite/composite.java
package test;

import java.util.ArrayList;
import java.util.List;

interface Shape {
	public void print();
}

class Square implements Shape {
	@Override
	public void print() {
		System.out.println("I'm Square");
	}	
}

class Circle implements Shape {
	@Override
	public void print() {
		System.out.println("I'm Circle");
	}
}

class Triangle implements Shape {
	@Override
	public void print() {
		System.out.println("I'm Triangle");
	}
}

class Point implements Shape {
	@Override
	public void print() {
		System.out.println("I'm Point");
	}
}

class Composite implements Shape {
	private List<Shape> components = new ArrayList<Shape>();
	@Override
	public void print() {
		for (Shape shape : components) {
			shape.print();
		}
	}
	
	public void addComponent(Shape c) {
		components.add(c);
	}
	
	public void removeComponent(Shape c) {
		components.remove(c);
	}
}

public class CompositeTest {

	public static void main(String[] args) {
	       //Initialize four ellipses
		Shape sh1 = new Square();
		Shape sh2 = new Circle();
		Shape sh3 = new Triangle();
		Shape sh4 = new Point();

        //Initialize three composite graphics
		Composite c1 = new Composite();
		Composite c2 = new Composite();
		Composite c3 = new Composite();

        //Composes the graphics
        c2.addComponent(sh1);
        c2.addComponent(sh2);
        c2.addComponent(sh3);

        c3.addComponent(sh4);

        c1.addComponent(c2);
        c1.addComponent(c3);

        //Prints the complete graphic (four times the string "Ellipse").
        c1.print();
	}
}
